### ISSUE TITLE
Crash on tilde separators in cache

### ISSUE THREAD

**Issue by cramertj** [bazelbuild/bazel#23322]
### Description of the bug:

When attempting to update [Pigweed](https://pigweed.googlesource.com/pigweed/pigweed/) from `8.0.0-pre.20240618.2` to `8.0.0-pre.20240812.1` we hit the following error when building any target (e.g. `bazelisk build //pw_chrono:system_clock`):

```
Starting local Bazel server and connecting to it...
Computing main repo mapping:
    Fetching https://bcr.bazel.build/modules/stardoc/0.7.0/source.json
FATAL: bazel crashed due to an internal error. Printing stack trace:
java.lang.RuntimeException: Unrecoverable error while evaluating node 'REPOSITORY_DIRECTORY:@@fuchsia_sdk' (requested by nodes 'PACKAGE_LOOKUP:@@fuchsia_sdk//fuchsia')
	at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:547)
	at com.google.devtools.build.lib.concurrent.AbstractQueueVisitor$WrappedRunnable.run(AbstractQueueVisitor.java:435)
	at java.base/java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool.scan(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(Unknown Source)
Caused by: java.util.concurrent.CompletionException: com.google.devtools.build.lib.cmdline.LabelSyntaxException: invalid repository name '_main~_repo_rules~cipd_client': repo names may contain only A-Z, a-z, 0-9, '-', '_', '.' and '+'
	at com.github.benmanes.caffeine.cache.LocalLoadingCache.lambda$newMappingFunction$3(LocalLoadingCache.java:204)
	at com.github.benmanes.caffeine.cache.BoundedLocalCache.lambda$doComputeIfAbsent$13(BoundedLocalCache.java:2451)
	at java.base/java.util.concurrent.ConcurrentHashMap.compute(Unknown Source)
	at com.github.benmanes.caffeine.cache.BoundedLocalCache.doComputeIfAbsent(BoundedLocalCache.java:2449)
	at com.github.benmanes.caffeine.cache.BoundedLocalCache.computeIfAbsent(BoundedLocalCache.java:2432)
	at com.github.benmanes.caffeine.cache.LocalCache.computeIfAbsent(LocalCache.java:107)
	at com.github.benmanes.caffeine.cache.LocalLoadingCache.get(LocalLoadingCache.java:57)
	at com.google.devtools.build.lib.cmdline.RepositoryName.createUnvalidated(RepositoryName.java:98)
	at com.google.devtools.build.lib.rules.repository.RepoRecordedInput$RepoCacheFriendlyPath.parse(RepoRecordedInput.java:221)
	at com.google.devtools.build.lib.rules.repository.RepoRecordedInput$File$1.parse(RepoRecordedInput.java:267)
	at com.google.devtools.build.lib.rules.repository.RepoRecordedInput.parse(RepoRecordedInput.java:103)
	at com.google.devtools.build.lib.rules.repository.RepositoryDelegatorFunction$DigestWriter.readMarkerFile(RepositoryDelegatorFunction.java:712)
	at com.google.devtools.build.lib.rules.repository.RepositoryDelegatorFunction$DigestWriter.areRepositoryAndMarkerFileConsistent(RepositoryDelegatorFunction.java:675)
	at com.google.devtools.build.lib.rules.repository.RepositoryDelegatorFunction$DigestWriter.areRepositoryAndMarkerFileConsistent(RepositoryDelegatorFunction.java:649)
	at com.google.devtools.build.lib.rules.repository.RepositoryDelegatorFunction.compute(RepositoryDelegatorFunction.java:199)
	at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:467)
	... 7 more
Caused by: com.google.devtools.build.lib.cmdline.LabelSyntaxException: invalid repository name '_main~_repo_rules~cipd_client': repo names may contain only A-Z, a-z, 0-9, '-', '_', '.' and '+'
	at com.google.devtools.build.lib.cmdline.LabelParser.syntaxErrorf(LabelParser.java:208)
	at com.google.devtools.build.lib.cmdline.RepositoryName.validate(RepositoryName.java:177)
	at com.google.devtools.build.lib.cmdline.RepositoryName.lambda$static$0(RepositoryName.java:68)
	at com.github.benmanes.caffeine.cache.LocalLoadingCache.lambda$newMappingFunction$3(LocalLoadingCache.java:197)
	... 22 more
```

A bisect points at https://github.com/bazelbuild/bazel/pull/23103. It seems like there was an intentional choice to change the separator. Deleting `MODULE.bazel.lock` does not solve the issue, ~nor does `bazelisk clean --expunge`~.

### Which category does this issue belong to?

Core

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

Clone https://pigweed.googlesource.com/pigweed/pigweed/
Update `.bazelversion` to `8.0.0-pre.20240812.1`
`bazelisk build //pw_chrono:system_clock` (or any other target)

**Comment by Wyverald** [comment]
Thanks for the report!

This is pretty bad, but I _think_ it should be fine for people on 7.3.1 as `@@abc~` is still a valid repo name on that version, even if you flip the flag. IOW, this crash should only show up for people on rolling.

The immediate fix is to run `bazel clean --expunge`; I'll send a proper fix for this.

**Comment by cramertj** [comment]
FYI `bazel clean --expunge` did *not* fix this issue for me-- thanks for sending a quick fix, though!

**Comment by Wyverald** [comment]
Wait, that's surprising. Do you get the exact same error message? I don't understand how that's possible, then.

**Comment by cramertj** [comment]
@Wyverald My apologies-- i'm not sure what I did before when I tested. I must've early-aborted the expunge or something. I tried again just now, and building after `bazelisk clean --expunge` worked.

**Comment by iancha1992** [comment]
A fix for this issue has been included in [Bazel 7.4.0 RC1](https://github.com/bazelbuild/bazel/releases/tag/7.4.0rc1). Please test out the release candidate and report any issues as soon as possible.
If you're using Bazelisk, you can point to the latest RC by setting `USE_BAZEL_VERSION=7.4.0rc1`. Thanks!
